Add to Cart - stepper style: fix input field when the product stock quantity is 1

gigitux opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

<!-- For bug fixes: If known, please provide links to help with traceability and escape analysis. -->
<!-- Please include a link to the issue of the bug being fixed, if one doesn't exist please create it. -->
<!-- If the PR that introduced the bug is known, please also add its link below. -->

Closes https://github.com/woocommerce/woocommerce/issues/57454 (WOOPLUG-4034) .

(For Bug Fixes) Bug introduced in PR # .

### Screenshots or screen recordings:

<!-- If this PR includes UI changes, please provide screenshots or a screen recording for clarity. -->
<!-- This section can be removed if not applicable. -->

| Before | After |
| ------ | ----- |
| ![image](https://github.com/user-attachments/assets/d54c0660-f9e9-4be3-89f8-5ad3432dc01c)| ![image](https://github.com/user-attachments/assets/ea379ba6-34d2-498d-b4c0-4cbf150d840c)|


<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Edit the Single Product Template
2. Edit the `Add to Cart with Options`.
3. Switch the quantity selector to stepper.
4. Save the template.
5. Edit a product.
6. Click on Inventory.
7. Click on `Track stock quantity for this product`
8. Update quantity to 1.
9. Save.
10. Visit the product.
11. Ensure that the stepper buttons aren't rendered.

<!-- End testing instructions -->

### Testing that has already taken place:

<!-- Detail any testing that has already been conducted. -->
<!-- Include environment details such as hosting type, plugins, theme, store size, store age, and relevant settings. -->
<!-- Mention any analysis performed, such as assessing potential impacts on environment attributes and other plugins, performance profiling, or LLM/AI-based analysis. -->
<!-- Within the testing details you provide, please ensure that no sensitive information (such as API keys, passwords, user data, etc.) is included in this public pull request. -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [x] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [x] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

Add to Cart - stepper style: fix input field when the product stock quantity is 1

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

gigitux committed:
Add to Cart - stepper style: fix input field when the product stock quantity is 1

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

gigitux committed:
Merge branch 'trunk' into wooplug-4034-stepper-left-qty-1-display-issue-only-input-field-is-hidden

Aljullu reviewed (commented):
Nice fix! I tested setting a product's inventory to 1 and confirming the stepper buttons aren't shown unnecessarily and it works correctly

Before | After
--- | ---
![imatge](https://github.com/user-attachments/assets/a862fe1c-8bdd-4194-85b1-8f0098fbc6ad) | ![imatge](https://github.com/user-attachments/assets/63b74ab1-ed45-44ed-92a3-3b1934f7ba81)


@gigitux I noticed the issue also affects the _Quantity Selector_ block of the _Add to Cart with Options_ block. So while not a blocker of this PR, I was wondering if we should fix it there as well? I think adding the same stock check [here](https://github.com/woocommerce/woocommerce/blob/7edde48eaf24e719ade0c1fcb1446d0aaf06c019/plugins/woocommerce/src/Blocks/BlockTypes/AddToCartWithOptionsQuantitySelector.php#L133) would be enough.

Update: also, not a blocker, but I think it would be nice to get automated tests for this.

gigitux committed:
fix logic

gigitux committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce into wooplug-4034-stepper-left-qty-1-display-issue-only-input-field-is-hidden

gigitux committed:
add E2E test

gigitux committed:
Merge branch 'wooplug-4034-stepper-left-qty-1-display-issue-only-input-field-is-hidden' of github.com:woocommerce/woocommerce into wooplug-4034-stepper-left-qty-1-display-issue-only-input-field-is-hidden

gigitux committed:
keep the same order

Aljullu commented on plugins/woocommerce/src/Blocks/BlockTypes/AddToCartForm.php:
```diff
@@ -136,7 +136,13 @@ protected function render( $attributes, $content, $block ) {
+		/**
+		 * The stepper buttons don't show when the product is sold individually or stock quantity is less or equal to 1 because the quantity input field is hidden.
+		 */
+		$is_stepper_style = 'stepper' === $attributes['quantitySelectorStyle'] && ! $product->is_sold_individually() && ( ( $managing_stock && $stock_quantity > 1 ) || ! $managing_stock ) && Features::is_enabled( 'add-to-cart-with-options-stepper-layout' );
```
Ups, sorry I missed in the first review, but in the _Add to Cart with Options_ block there will only be the stepper style (https://github.com/woocommerce/woocommerce/issues/56686), so this variable will be removed. I think it would be easier to simply tweak the _Quantity Selector_ block and [return early](https://github.com/woocommerce/woocommerce/blob/7edde48eaf24e719ade0c1fcb1446d0aaf06c019/plugins/woocommerce/src/Blocks/BlockTypes/AddToCartWithOptionsQuantitySelector.php#L135) when there is only one item in stock?

Aljullu reviewed (commented):
Thanks for the updates!

> Done with 42ad135. I wasn't able to find specific E2E tests for `add-to-cart-with-options-quantity-selector`. Let me know if I missed them, or I should add this test in another file to cover the Blockified Add To cart block as well.

Right, we don't have specific tests for the inner blocks, we test them as part of the _Add to Cart with Options_ wrapper. In this specific case, I think the [PHP unit tests](https://github.com/woocommerce/woocommerce/blob/b49c1b3e9e77e742e604f368fe8c685516b5de07/plugins/woocommerce/tests/php/src/Blocks/BlockTypes/AddToCartWithOptions.php) might be useful as we have some similar tests there.

I'm happy to work on it in a follow-up if you prefer. Or if you can include it in this PR, that would be awesome! :100:

gigitux committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce into wooplug-4034-stepper-left-qty-1-display-issue-only-input-field-is-hidden

gigitux committed:
improve logic and add unit test

gigitux commented:
No worries, @Aljullu! I fixed the code and added unit test.

I believe you’re already aware, but since the new block only supports the stepper style, it seems we can simplify the logic by removing some of the checks:


https://github.com/woocommerce/woocommerce/blob/d77461822512d9f3dee97c0e5d7635beddbe8037/plugins/woocommerce/src/Blocks/BlockTypes/AddToCartWithOptionsQuantitySelector.php#L174-L177


https://github.com/woocommerce/woocommerce/blob/d77461822512d9f3dee97c0e5d7635beddbe8037/plugins/woocommerce/src/Blocks/BlockTypes/AddToCartWithOptionsQuantitySelector.php#L201-L202


I didn’t update this in the current PR to keep the scope focused. Let me know if that sounds good!

Aljullu commented on plugins/woocommerce/tests/php/src/Blocks/BlockTypes/AddToCartWithOptions.php:
```diff
@@ -190,4 +190,29 @@ public function test_grouped_product_selector_cta() {
+		$simple_product_id = $simple_product->save();
+
+		$markup = do_blocks( '<!-- wp:woocommerce/single-product {"productId":' . $simple_product_id . '} --><!-- wp:woocommerce/add-to-cart-with-options /--><!-- /wp:woocommerce/single-product -->' );
+		$this->assertStringNotContainsString( 'data-block-name="woocommerce/add-to-cart-with-options-quantity-selector"', $markup, 'The Add to Cart with Options form does not contains a quantity selector block for sold individually products.' );
```
```suggestion
		$this->assertStringNotContainsString( 'data-block-name="woocommerce/add-to-cart-with-options-quantity-selector"', $markup, 'The Add to Cart with Options form does not contain a quantity selector block for sold individually products.' );
```

Aljullu commented on plugins/woocommerce/tests/php/src/Blocks/BlockTypes/AddToCartWithOptions.php:
```diff
@@ -190,4 +190,29 @@ public function test_grouped_product_selector_cta() {
+		$simple_product->set_stock_quantity( 1 );
+		$simple_product->save();
+		$markup = do_blocks( '<!-- wp:woocommerce/single-product {"productId":' . $simple_product_id . '} --><!-- wp:woocommerce/add-to-cart-with-options /--><!-- /wp:woocommerce/single-product -->' );
+		$this->assertStringNotContainsString( 'data-block-name="woocommerce/add-to-cart-with-options-quantity-selector"', $markup, 'The Add to Cart with Options form does not contains a quantity selector block for products with manage stock set to true and stock quantity set to 1.' );
```
```suggestion
		$this->assertStringNotContainsString( 'data-block-name="woocommerce/add-to-cart-with-options-quantity-selector"', $markup, 'The Add to Cart with Options form does not contain a quantity selector block for products with manage stock set to true and stock quantity set to 1.' );
```

Aljullu reviewed (approved):
Thanks for the updates, it looks good to me, @gigitux! I left a couple of suggestions to fix typos, but that's all.

Besides testing the current block, I also tested the new blockified _Add to Cart with Options_ block and confirming the issue is fixed there too:

`trunk` | This branch
--- | ---
![imatge](https://github.com/user-attachments/assets/757ebf50-020a-4312-9a79-f3a03cbfd2c0) | ![imatge](https://github.com/user-attachments/assets/c924c234-faac-4e0a-bb81-a2f6699a0465)

> I believe you’re already aware, but since the new block only supports the stepper style, it seems we can simplify the logic by removing some of the checks:

Yes, this is being cleaned up in #57442. :slightly_smiling_face:

gigitux committed:
Update plugins/woocommerce/tests/php/src/Blocks/BlockTypes/AddToCartWithOptions.php

Co-authored-by: Albert Juhé Lluveras <contact@albertjuhe.com>

gigitux committed:
Update plugins/woocommerce/tests/php/src/Blocks/BlockTypes/AddToCartWithOptions.php

Co-authored-by: Albert Juhé Lluveras <contact@albertjuhe.com>

gigitux committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce into wooplug-4034-stepper-left-qty-1-display-issue-only-input-field-is-hidden

gigitux committed:
Merge branch 'trunk' into wooplug-4034-stepper-left-qty-1-display-issue-only-input-field-is-hidden
